### PR TITLE
 For #9488: Add Search widget CFR & telemetry

### PR DESCRIPTION
**Pull request by sblatz**
## Demo

image

image

demo of search widget CFR 2020-05-27 11_42_24

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by sblatz** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) **What questions will you answer with this data?**
- How often are users shown the CFR for search widget?
- How do users interact with the CFR for search widget (dismiss, cancel, add widget)?
- Are users who are given the CFR for search widget more likely to install it?

2) **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**
- These are being used for a bigger experiment around retention. Our goal is to see if using a search widget CFR can drive users to install a search widget, and ultimately retain better because of it.

3) **What alternative methods did you consider to answer these questions? Why were they not sufficient?**

* This CFR is new so we do not have any metrics to track it. The `search_widget_installed` value is now getting sent in the Glean metrics ping so we can access it through the experimentation framework (instead of just Leanplum).

4) **Can current instrumentation answer these questions?**
* N/A, this is a new feature.

5) **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data c](https://wiki.mozilla.org/Firefox/Data_Collection)[ollection ](https://wiki.mozilla.org/Firefox/Data_Collection)[categories](https://wiki.mozilla.org/Firefox/Data_Collection) on the found on the Mozilla wiki.**

* All data is Category 2.

6) **How long will this data be collected?**

* Until 09/01/2020

7) **What populations will you measure?**

* All release, beta, and nightly users with telemetry enabled.

8) **Please provide a general description of how you will analyze this data.**

* Glean / Amplitude

9) **Where do you intend to share the results of your analysis?**

* Only on Glean, Amplitude and with mobile teams.

File: app/src/main/java/org/mozilla/fenix/components/SearchWidgetCreator.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Handles the creation of the pinning search widget dialog.
+ */
+object SearchWidgetCreator {
+
+    /**
+     * Attempts to display a prompt requesting the user pin the search widget
+     * Returns true if the prompt is displayed successfully, and false otherwise.
+     */
+    @TargetApi(Build.VERSION_CODES.O)
```
[the remark this answers is not in the document]
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/components/SearchWidgetCreator.kt:23]: chatted about this offline!

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -36,14 +36,22 @@
...
     <!-- Text for the main message -->
     <string name="cfr_message">Add a shortcut to open private tabs from your Home screen.</string>
     <!-- Text for the positive button -->
     <string name="cfr_pos_button_text">Add shortcut</string>
     <!-- Text for the negative button -->
     <string name="cfr_neg_button_text">No thanks</string>
 
+    <!-- Search widget "contextual feature recommendation" (CFR) -->
+    <!-- Text for the main message. Placeholder text replaced with app name. -->
+    <string name="search_widget_cfr_message">Get to Firefox faster. Add a widget to your Home screen.</string>
```
[the remark this answers is not in the document]
Comment (sblatz) [reply app/src/main/res/values/strings.xml:49]: Good question. I'm actually not sure and forgot to put a "TODO" on this. I think I need an answer from @brampitoyo on whether or not this string should always say "Firefox" or not.

File: app/src/main/java/org/mozilla/fenix/components/SearchWidgetCreator.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    fun createSearchWidget(context: Context): Boolean {
+        val appWidgetManager: AppWidgetManager = context.getSystemService(AppWidgetManager::class.java)
+        if (!appWidgetManager.isRequestPinAppWidgetSupported) { return false }
+
+        val myProvider = ComponentName(context, SearchWidgetProvider::class.java)
+        appWidgetManager.requestPinAppWidget(myProvider, null, null)
```
[the remark this answers is not in the document]
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/components/SearchWidgetCreator.kt:29]: Good call! That definitely cleans this up a bit.

File: app/src/main/res/values/strings.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (sblatz) [reply app/src/main/res/values/strings.xml:49]: Got a thumbs up from Betsy that we can go ahead with just "firefox"

**Comment by liuche** [comment]
Note here about running this experiment - this is going to hit up against some of our other experimentation, so it may delay our EOL of the various Nightly channels, since it needs 3 weeks to run. Our next release is 5.2 and will go out next week.

**Pull request merged**